tests/unit_tests/epee_http_server.cpp

```diff
@@ -0,0 +1,192 @@
+    {
+      BEGIN_KV_SERIALIZE_MAP()
+        KV_SERIALIZE(payload)
+      END_KV_SERIALIZE_MAP()
+
+      std::string payload;
+    };
+  };
+
+  std::string make_payload()
+  {
+    dummy::request body{};
+    const auto body_serialized = epee::serialization::store_t_to_binary(body);
+    return std::string{
+      reinterpret_cast<const char*>(body_serialized.data()),
+      body_serialized.size()
+    };
+  }
+
+  struct http_server :  epee::http_server_impl_base<http_server>
+  {
+    using connection_context =  epee::net_utils::connection_context_base;
+
+    http_server()
+      : epee::http_server_impl_base<http_server>(),
+        dummy_size(payload_size)
+    {}
+
+    CHAIN_HTTP_TO_MAP2(connection_context); //forward http requests to uri map
+
+    BEGIN_URI_MAP2()
+      MAP_URI_AUTO_BIN2("/dummy", on_dummy, dummy)
+    END_URI_MAP2()
+
+    bool on_dummy(const dummy::request&, dummy::response& res, const connection_context *ctx = NULL)
+    {
+      res.payload.resize(dummy_size.load(), 'f');
+      return true;
+    }
+
+    std::atomic<std::size_t> dummy_size;
+  };
+} // anonymous
+
+TEST(http_server, response_soft_limit)
+{
+  namespace http = boost::beast::http;
+
+  http_server server{};
+  server.init(nullptr, "8080");
+  server.run(1, false);
+
+  boost::system::error_code error{};
+  boost::asio::io_context context{};
+  boost::asio::ip::tcp::socket stream{context};
+  stream.connect(
+    boost::asio::ip::tcp::endpoint{
+      boost::asio::ip::make_address("127.0.0.1"), 8080
+    },
+    error
+  );
+  EXPECT_FALSE(bool(error));
+
+  http::request<http::string_body> req{http::verb::get, "/dummy", 11};
+  req.set(http::field::host, "127.0.0.1");
+  req.set(http::field::user_agent, BOOST_BEAST_VERSION_STRING);
+  req.body() = make_payload();
+  req.prepare_payload();
+  http::write(stream, req, error);
+  EXPECT_FALSE(bool(error));
+
+  {
+    dummy::response payload{};
+    boost::beast::flat_buffer buffer;
+    http::response<http::basic_string_body<char>> res;
+    http::read(stream, buffer, res, error);
+    EXPECT_FALSE(bool(error));
+    EXPECT_EQ(200u, res.result_int());
+    EXPECT_TRUE(epee::serialization::load_t_from_binary(payload, res.body()));
+    EXPECT_EQ(payload_size, std::count(payload.payload.begin(), payload.payload.end(), 'f'));
+  }
+
+  while (!error)
+    http::write(stream, req, error);
+  server.send_stop_signal();
+}
+
+TEST(http_server, private_ip_limit)
+{
+  namespace http = boost::beast::http;
+
+  http_server server{};
+  server.dummy_size = 1;
+  server.init(nullptr, "8080");
+  server.run(1, false);
+
+  boost::system::error_code error{};
+  boost::asio::io_context context{};
+
+  std::vector<boost::asio::ip::tcp::socket> streams{};
+  for (std::size_t i = 0; i < max_private_ips; ++i)
+  {
+    streams.emplace_back(context);
+    streams.back().connect(
+      boost::asio::ip::tcp::endpoint{
+        boost::asio::ip::make_address("127.0.0.1"), 8080
+      },
+      error
+    );
+  }
+
```

Comment:
You may want an additional test for the negative case, i.e. requests work fine up until the limit.

Reply:
Added some read/writes in the loop to ensure all of the connections work, before the 1+ case fails.